airbyte-webapp/src/pages/connections/CreateConnectionPage/usePreloadData.ts

```diff
@@ -0,0 +1,113 @@
+import { useEffect } from "react";
+import { useLocation, useNavigate, useSearchParams } from "react-router-dom";
+
+import {
+  DestinationDefinitionRead,
+  DestinationRead,
+  SourceDefinitionRead,
+  SourceRead,
+} from "core/request/AirbyteClient";
+import { useGetDestination } from "hooks/services/useDestinationHook";
+import { useGetSource } from "hooks/services/useSourceHook";
+import { useDestinationDefinition } from "services/connector/DestinationDefinitionService";
+import { useSourceDefinition } from "services/connector/SourceDefinitionService";
+
+export function hasSourceId(state: unknown): state is { sourceId: string } {
+  return typeof state === "object" && state !== null && typeof (state as { sourceId?: string }).sourceId === "string";
```

Comment:
An `any` might actually be best in terms of readability.
`return typeof state?.sourceId === "string"`

Reply:
I did actually try simplifying these type guard functions (this comment goes equally for `hasDestinationId`: all of the sensible simplifications just made the compiler yell at me. This appears to be a more or less "correct" type guard for a built-in web API type :weary:

Reply:
Perhaps `Record<string, string>`?

Reply:
@krishnaglick That actually doesn't work, because `state` is `location.state`, and ~~robocop~~ typescript doesn't let you run around retyping built-in web APIs.